RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1470 @@
+                                      nMaxItersNNLS,
+                                      10,
+                                      10);
+
+#ifdef HCAL_MAHI_GPUDEBUG
+                printf("result Amplitudes\n");
+                for (int i = 0; i < 8; i++)
+                  printf("resultAmplitudes(%d) = %f\n", i, resultAmplitudesVector(i));
+#endif
+
+                calo::multifit::calculateChiSq(
+                    matrixL, glbPulseMatrixView, resultAmplitudesVector, inputAmplitudesView, chi2);
+
+                auto const deltaChi2 = std::abs(chi2 - previous_chi2);
+                if (chi2 == chi2_2itersback && chi2 < previous_chi2)
+                  break;
+
+                // update
+                chi2_2itersback = previous_chi2;
+                previous_chi2 = chi2;
+
+                // exit condition
+                if (deltaChi2 < deltaChi2Threashold)
+                  break;
+              }
+
+#ifdef HCAL_MAHI_GPUDEBUG
+              for (int i = 0; i < NPULSES; i++)
+                printf("pulseOffsets(%d) = %d outputAmplitudes(%d) = %f\n",
+                       i,
+                       pulseOffsets(i),
+                       i,
+                       resultAmplitudesVector(i));
+              printf("chi2 = %f\n", chi2);
+#endif
+
+              outputGPU.chi2()[gch] = chi2;
+              auto const idx_for_energy = std::abs(pulseOffsetsView.offsets()[0]);
+              outputGPU.energy()[gch] = (gain * resultAmplitudesVector(idx_for_energy)) * respCorrection;
+
+            }  // loop over channels
+          }    //loop over group of channels
+        }
+      };  // Kernel_minimize
+
+    }  // namespace mahi
+  }    // namespace hcal::reconstruction
+  namespace hcal {
+    namespace reconstruction {
+
+      void entryPoint(Queue& queue,
+                      IProductTypef01::ConstView const& f01HEDigis,
+                      IProductTypef5::ConstView const& f5HBDigis,
+                      IProductTypef3::ConstView const& f3HBDigis,
+                      OProductType::View outputGPU,
+                      HcalMahiConditionsPortableDevice::ConstView const& mahi,
+                      HcalSiPMCharacteristicsPortableDevice::ConstView const& sipmCharacteristics,
+                      HcalRecoParamWithPulseShapeDevice::ConstView const& recoParamsWithPS,
+                      HcalMahiPulseOffsetsPortableDevice::ConstView const& mahiPulseOffsets,
+                      ConfigParameters const& configParameters) {
+        auto const totalChannels =
+            f01HEDigis.metadata().size() + f5HBDigis.metadata().size() + f3HBDigis.metadata().size();
+        // FIXME: the number of channels in output might change given that some channesl might be filtered out
+
+        // do not run when there are no rechits (e.g. if HCAL is not being read),
+        // but do set the size of the output collection to 0
+        if (totalChannels == 0) {
+          return;
+        }
```

Comment:
The comment does not seem to match the code ?

Reply:
Implemented this behaviour in `HBHERecHitProducerPortable.cc` https://github.com/cms-sw/cmssw/pull/44910/files#diff-b7f6c20e7b8859eef3ad3a13dde17cbdc45adfacea95d76c0410585baa784067R138-R141

Will remove these lines.

Reply:
Done in 11aa37463e89e26d17c31a9e9d76e6f799b8a298